### ISSUE TITLE
RTOS2 : question about critical region

### ISSUE THREAD

**Issue by gerhil** [ARM-software/CMSIS_5#171]
Why isn't it allowed to use osThreadNew(), osThreadSuspend() and osThreadResume() in critical regions? See rtx_thread.c lines 1505, 1589 and 1599.

**Comment by 0xc0170** [comment]
@Wazalogic have you tried it to run with without these conditions? As I recall it hard-faults, as it tries to makes svc calls.

**Comment by JonatanAntoni** [comment]
In case of RTX5 we use SVC calls to implement this management functions, yes. Thus interrupts must not be disabled. Can you explain your use case for placing osThreadXXX calls in ciritical regions, i.e. having interrupts globally disabled?

**Comment by ilg-ul** [comment]
where are documented the critical sections in CMSIS OS2? I asked for it many times in the past, but I don't remember receiving an answer.

generally speaking, the implementation of critical sections does not necessarily imply having interrupts globally disabled; on v7m devices you can set the PRIMASK.

any RTOS must have clear rules on how to enter/exit interrupts critical sections, and the mechanism must allow unlimited nesting; similar to scheduler critical sections, the safest implementation is to read the current status and save it to a variable, disable interrupts or change PRIMASK, do the critical code, and restore the saved status.

[CMSIS++](http://micro-os-plus.github.io/cmsis-plus/) has clearly defined functions for this:

http://micro-os-plus.github.io/reference/cmsis-plus/group__cmsis-plus-rtos-c-core.html#gaeb1888b2cebfa7d7beb7b1590633033a

**Comment by Wazalogic** [comment]
In my project I need to perform context switching between specific threads as follows 

core_util_critical_section_enter();
osThreadResume(newThread);
osThreadSuspend(otherThread);
core_util_critical_section_exit();

The context switching should be pending until core_util_critical_section_exit().

The interrupt serivce routine can wake up a thread, using osThreadResume(thread), and context switching should not happen if the program is in a critical region.

**Comment by Wazalogic** [comment]
> any RTOS must have clear rules on how to enter/exit interrupts critical sections, and the mechanism must allow unlimited nesting; similar to scheduler critical sections, the safest implementation is to read the current status and save it to a variable, disable interrupts or change PRIMASK, do the critical code, and restore the saved status.

I agree. FreeRTOS also supports this behaviour.

**Comment by Wazalogic** [comment]
I understand that this is limited by the SVC call implementation. However, from a design perspective this limitation is a design flaw. The semantics of the API functions should allow these functions in a critical region. Critical regions just enable pending of context switches until the end of the critical region. For example, FreeRTOS implements this semantics and it gives good control of threads. This is what multithreading is about! The fact that critical regions, osThreadSuspend and osThreadResume are very dangereous to use, should be no reason to exclude or limit them.

**Comment by ilg-ul** [comment]
> FreeRTOS also supports this behaviour.

yeah, just that the implementation is suboptimal, it uses a counter, which does not play well at all if interrupts were disabled outside their RTOS (`exit()` will see the counter to be zero, and prematurely enable the interrupts).

as I said, the only absolutely safe implementation is to save the status on `enter()` and restore it on `exit()` (as CMSIS++ does).

**Comment by ilg-ul** [comment]
> Critical regions just enable pending of context switches until the end of the critical region.

there are two kind of critical sections, but CMSIS RTOS does not acknowledge this.

there are scheduler critical sections, which prevent context switches, and interrupt critical sections, which prevent interrupts with lower priority.

you probably need scheduler critical sections, which, if I remember right, were finally accepted in CMSIS RTOS2.

**Comment by Wazalogic** [comment]
Yes. I can probably use any of them because I keep the critical sections very very small.

**Comment by Wazalogic** [comment]
@0xc0170  I have disabled the critical section and the code runs ok. However, I need the critical section to prevent race conditions.

**Comment by ilg-ul** [comment]
despite their name, scheduler critical sections aren't that _critical_, at least not as critical as interrupt critical sections. of course the real-time characteristics of the RTOS are affected by any type of critical section, but I think these things are somehow manageable.

**Comment by ReinhardKeil** [comment]
Hi gerhil,

can you provide a bit more information about this code section.  What is behind the functions:
core_util_critical_section_enter() and core_util_critical_section_exit().  Can you provide the source code of that functions?

Why is it important to resume context switch when you are in a critical section.  I would expect that the opposite is important?  I would assume that you do not want any context switches in a critical section.

Why does such a sequence not work?

```
osThreadSuspend(myThread);
core_util_critical_section_enter();
   // execute stuff that should not get interrupted by thread switches
core_util_critical_section_exit();
osThreadResume(myThread);
```

**Comment by ilg-ul** [comment]
> What is behind the functions:
core_util_critical_section_enter() and core_util_critical_section_exit(). Can you provide the source code of that functions?

this is a good example why the code for interrupt critical sections should be in the RTOS public API, because applications need it, and not having it in the API leads to re-implementations in the application, most of the time inconsistent with the system.

**Comment by gerhil** [comment]
Hi Reinhard,
I came from mbed. I used their code base (feature_cmsis5) which is derived from ARM’S cmsis5. They couldn’t help me and therefore they sent me here.

The functions core_util_critical_section_enter() and core_util_critical_section_exit() come from mbed_critical.c. Sorry that I didn’t mentioned it. These functions use __disable_irq() and __enable_irq() from rt_HAL_CM.h. I have tried these functions as well and I got same problem.

The critical region should disable context switching. This is what I would expect. In my case, the osThreadResume() and osThreadSuspend() must be performed in sequence without context switching in between. Also, the osThreadResume() and osThreadSuspend() must not yield in a critical region.  If this is not the case then a context switch after osThreadResume() and before osThreadSuspend() could allow another thread (e.g. the one being resumed) to resume this thread that has not been suspended yet. This is a race condition causing deadlock. FYI, the sequence ‘osThreadSuspend();osThreadResume()’ is also a possible solution, but this may perform unnecessary overhead of going through the idle state. In a critcal region both sequences shouldn't matter.

I looked at osThreadGetPriority(). It seems to me that you cannot call osThreadGetPriority() in a critical region. Why not? Isn’t that odd? I am missing sound semantics in the API of cmsis5.

**Comment by ReinhardKeil** [comment]
Thanks for the clarification - that was important information for us.

For disabling thread switching, you may use osKernelLock / os KernelUnlock.  See also http://arm-software.github.io/CMSIS_5/RTOS2/html/group__CMSIS__RTOS__KernelCtrl.html#ga948609ee930d9b38336b9e1c2a4dfe12

This does not alter the interrupt system and should be sufficient the for majority of use cases.

**Comment by gerhil** [comment]
Thanks for pointing out to osKernelLock/osKernelUnlock. However, this does not work with osThreadResume() and osThreadSuspend() as expected.

Is it better to use osEventFlagsSet() and osEventFlagsWait() instead of resp. osThreadResume() and osThreadSuspend()? How does this behave within osKernelLock/osKernelUnlock?

**Comment by ReinhardKeil** [comment]
Here is a bit more background:

As you perhaps know, RTX5 is designed to be extended with MPU protection.  As such RTX functions run in handler mode (called via SVC). In an MPU protected system, most threads should run in unprivileged mode and directly enabling/disabling IRQ is impossible.  We are in the process get a safety certification for RTX5 and therefore we are reluctant to add RTOS function that enable/disable all IRQ as this introduces also safety risk. Incorrect handling may result in a faulty system since interrupt response times become unpredictable in other parts of the application.

Using osKernelLock and osKernelUnlock allows to disable thread switching and is usually what is needed. But I'm open to get more information of real use cases.

While IRQs are disabled, the SVC calls to RTOS are also blocked. For important RTOS functions that should be callable from IRQ functions, a special handling is implemented, however this adds overhead.  We believe that it should be not required to call osThreadGetPriority from IRQ, but again I'm open for feedback to hear about potential use cases.

In general, a good design approach should minimize the code that is executed in the IRQ handler itself and the RTOS helps you to do that. Therefore disabling all interrupts in a system (except in the system startup phase) should be never required.

**Comment by gerhil** [comment]
Thanks again. I see, I hit the protection mechanism. I agree, we should be very careful with functions like `__disable_irq()` and ISRs must be short. How can I perform my code (context switching) in unprivileged mode or can I switch to privileged mode?

Consider the following example. Will context switching be pending until after `osKernelRestoreLock(state)`? If so, then this could solve my problem.
```
uint32_t` state = osKernelLock();
osEventFlagsSet(…);
osEventFlagsWait(…);
osKernelRestoreLock(state);
```
BTW: I am not using an `osThreadGetPriority()` in an ISR. I wonder if this protection is necessary for all functions.

**Comment by JonatanAntoni** [comment]
Hi @gerhil, I admit the documentation is osThreadResume() and osThreadSuspend() is not as the best. What did you expect? And what did you get? I made a small Blinky example with two threads and tried to alter between them using

```osKernelLock();
osThreadResume(tid_threadB);
osThreadSuspend(tid_threadA);
osKernelUnlock();```

and this works as I expect. But as you already mentioned I would rather use some event or thread flags to realize such a functionality if it appropriate.

**Comment by gerhil** [comment]
I am doing the same, but there seems to be a glitch that performs a yield inside the critical region. My code is too big to post here.

**Comment by ilg-ul** [comment]
> there seems to be a glitch that performs a yield inside the critical region

what glitch?

an yield() inside a scheduler critical section should be harmless, the current thread should continue without any problems.

on the other side, I think that an yield() inside an interrupt scheduler section should trigger an error.

**Comment by gerhil** [comment]
@ilg-ul  I mean yield performed by osThreadResume() osThreadSuspend() inside the critical section. 

Example will follow...

**Comment by gerhil** [comment]
```
osThreadId_t tid_thread1;
osThreadId_t tid_thread2;

void task1(void *parameter) {
    tid_thread1 = osThreadGetId();
    osThreadSuspend(tid_thread1);
    while(1) {
        printf ("Task1\r\n");
        fflush(stdout);
        osKernelLock();
        osThreadResume(tid_thread2);
        osThreadSuspend(tid_thread1);
        osKernelUnlock();
    }
}

void task2(void *parameter) {
    tid_thread2 = osThreadGetId();
    osThreadSuspend(tid_thread2);
    while(1) {
        printf ("Task2\r\n");
        fflush(stdout);
        osKernelLock();
        osThreadResume(tid_thread1);
        osThreadSuspend(tid_thread2);
        osKernelUnlock();
    }
}

int main() {
    os_thread_t cb_thread1;
    memset(&cb_thread1, 0, sizeof(cb_thread1));
    osThreadAttr_t thread_attr1;
    memset(&thread_attr1, 0, sizeof(thread_attr1));
    thread_attr1.name = "thread1";
    thread_attr1.stack_size = 72U + (256U*4U);
    thread_attr1.stack_mem = new uint8_t[thread_attr1.stack_size];
    thread_attr1.priority = osPriorityNormal;
    thread_attr1.cb_mem = &cb_thread1;
    thread_attr1.cb_size = sizeof(os_thread_t);

    os_thread_t cb_thread2;
    memset(&cb_thread2, 0, sizeof(cb_thread2));
    osThreadAttr_t thread_attr2;
    memset(&thread_attr2, 0, sizeof(thread_attr2));
    thread_attr2.name = "thread2";
    thread_attr2.stack_size = 72U + (256U*4U);
    thread_attr2.stack_mem = new uint8_t[thread_attr2.stack_size];
    thread_attr2.priority = osPriorityNormal;
    thread_attr2.cb_mem = &cb_thread2;
    thread_attr2.cb_size = sizeof(os_thread_t);


    //osKernelLock();
    tid_thread1 = osThreadNew((osThreadFunc_t)task1, NULL, &thread_attr1);
    if (tid_thread1 == NULL) {
        printf("Thread1 not created!\r\n");
    }

    tid_thread2 = osThreadNew((osThreadFunc_t)task2, NULL, &thread_attr2);
    if (tid_thread2 == NULL) {
        printf("Thread2 not created!\r\n");
    }
    osDelay(1000);
    osThreadResume(tid_thread1);

for(;;) { osDelay(1000); }
return 0;
}
```

**Comment by JonatanAntoni** [comment]
@gerhil Yes, i think your example reveals the "problem". You suspend the active thread while the scheduler is locked. Suspending a thread puts this thread to be blocked immediately. Thus osThreadSuspend(self) is not expected to return until one issues osThreadResume() for this thread. 

My humble opinion is you should not rely on such a design. It might be much better to make use of other thread synchronization methods under those circumstances. Did you try to solve this using thread or event flags, yet?

**Comment by gerhil** [comment]
This shows that osThreadSuspend() and osThreadResume() are incomplete. I wonder if event flags will be useful this way. What will osEventFlagsWait(…) do in a critical section? Same as osThreadSuspend()?

I agree. This is not a design I will ever be using in an application. I am using this technique for a new type of real-time kernel on top of cmsis.

Have you heard of “Communicating Sequential Processes” (CSP) by Tony Hoare and Bill Roscoe? It’s a theory of programming that provides true concurrency. It’s more than a process algebra. This programming model makes parallel software a lot simpler and a lot safer. Concurrency is not multithreading and it is not parallelism. Also, it does not mean “same time”. Concurrency is different.
I am developing a new programming model based on CSP for C++ (and other modern programming languages). I want to make this programming model available for the ARM processor. Currently, it works with FreeRTOS for ARM, but I want to make it useful with cmsis-rtos. To make this work with cmsis, I need this type of context switching to work.

**Comment by ilg-ul** [comment]
> This shows that osThreadSuspend() and osThreadResume() are incomplete

what makes them incomplete? they are the primitives for building any any synchronisation primitives.

I did not check the new CMSIS RTOS2 pages, perhaps they do not explain what these two functions are intended; suspend() is intended (or at least was intended, when I suggested them) to remove the current thread from the running list, and resume() was intended to add the given thread to the running list. 

so, what makes them incomplete?

> CSP

ok, this is one of the new programming models. I'm not sure it can be implemented on top of more classical APIs.

**Comment by gerhil** [comment]
> what makes them incomplete? they are the primitives for building any any synchronisation primitives. 

I mean by incomplete, their semantics in the presence of a critical section is incomplete. This discussion is also found at FreeRTOS. Pending context switching until the end of a critical section is a good solution and logical to the semantics of a critical section.

I believe that suspend and resume are thread control functions and not synchronization primitives. Suspending and resuming of another thread must not yield. They are administrative functions. Only suspending the current thread needs yielding (immediately or after releasing a critical section). Semaphores and monitors are synchronization primitives.

> ok, this is one of the new programming models. I'm not sure it can be implemented on top of more classical APIs.

I build CSP with Dijkstra’s semaphores and Hoare’s monitors. I only need a critical section and suspend and resume from the cmsis kernel. The programming model is called Waza. www.wazalogic.com

**Comment by ilg-ul** [comment]
> suspend and resume are thread control functions and not synchronization primitives. 

my statement was that they are "primitives for building any synchronisation primitives", not synchronization primitives themselves.

> their semantics in the presence of a critical section is incomplete

the semantics of scheduler critical sections is very simple and clear: 'no context switches are performed **inside** the  critical section'; if you call resume(), the thread will be added to the ready list, but, being inside the scheduler critical section, the context switch is ignored for now. however the thread was enqueued in the ready list and remains there. if you call suspend(), normally the current thread is suspended and execution continues to a thread chosen by the scheduler, but, in a scheduler critical section, the current thread is resumed immediately (suspending another thread than the current thread should not be allowed). I thought that this behaviour is obvious and expected. what would you add to make it "complete"?

I don't know the details of the RTX implementation, but this was the original expected behaviour, and I still think it is minimal and orthogonal.

BTW, in your example the sequence [ lock(), resume(), suspend(), unlock() ] makes not much sense; according to the above semantics, a single resume() would be enough.

**Comment by gerhil** [comment]
> my statement was that they are "primitives for building any synchronisation primitives", not synchronization primitives themselves.

Ah yes. Sorry you are right.

> the semantics of scheduler critical sections is very simple and clear: 'no context switches are performed ...

I would also except the behavior you describe. My example shows a different behavior. 

> BTW, in your example the sequence [ lock(), resume(), suspend(), unlock() ] makes not much sense; according to the above semantics, a single resume() would be enough.

Only one thread should perform. The example resumes another thread and suspends the current thread. All threads have equal priorities.

**Comment by ilg-ul** [comment]
> My example shows a different behavior.

then it should be a misunderstanding in the RTX implementation.

> The example resumes another thread and suspends the current thread. 

both these actions are performed by resume(). just try to replace in your example the scheduler critical sections with a single resume().

FYI, I did extensive studies and testing with various solutions while designing and implementing µOS++ / CMSIS++. actually the synchronisation objects in µOS++ are highly portable and can run on top of another scheduler, with the layered design of CMSIS++ using only these two suspend()/resume() functions (I have a functional version on top of the FreeRTOS scheduler).

**Comment by gerhil** [comment]
> then it should be a misunderstanding in the RTX implementation.

Yes, that is the problem here.

> both these actions are performed by resume(). just try to replace in your example the scheduler critical sections with a single resume().

Does this work with a preemptive scheduler underneath? 

The CSP kernel is preemptive and I want it compatible with the underlying (preemptive) kernel.

Ok, I have tested your suggestion and the output seems to be time-sliced or printf is interfering. I do not get the output I would expect.

**Comment by ilg-ul** [comment]
> or printf is interfering

btw, please be aware that newlib itself is not thread aware. using stdio functions in multi-threaded apps is a gamble.

**Comment by gerhil** [comment]
Why is this thread DONE?

**Comment by ilg-ul** [comment]
that's ARM way of telling you to not expect anything more from them.

**Comment by gerhil** [comment]
I am surprised and disappointed. I have to tell my customers not to used rtx2/cmsis5.

**Comment by ilg-ul** [comment]
I've already been there. when RTOS1 was out, I was also optimistic and recommended it to my customers. I even had a RTOS1 API for my µOS++. but it was hopeless, the RTOS1 API triggered many protests and I had to abandon it. 

RTOS2 is slightly better, but I still can't recommend it... :-(

(not to mention that in the mean time, given the choice, my customers discovered that the C API is somehow redundant and the µOS++ native C++ API is used now quite successfully).

**Comment by ReinhardKeil** [comment]
Gerald

I believe you did accidentally set the thread to DONE.

I was actually surprised by that as well. Please reopen it.

Reinhard
________________________________

**Comment by ReinhardKeil** [comment]
Correction: you perhaps accidentally CLOSED the ticket.
________________________________

**Comment by gerhil** [comment]
Reinhard

JonatanAntoni did set the thread on DONE.

**Comment by RobertRostohar** [comment]
@gerhil 

> the semantics of scheduler critical sections is very simple and clear: 'no context switches are performed ...
> 
> I would also except the behavior you describe. My example shows a different behavior.
> 
I have tried your example and it works as expected. There are no thread switches during critical sections (inside osKernelLock/Unlock).

**Comment by JonatanAntoni** [comment]
@gerhil I am sorry for assuming the issue is DONE because you closed it by accident.

**Comment by gerhil** [comment]
@JonatanAntoni I read in the forum said that you added DONE and removed review status. 

Thanks Jonatan.

@RobertRostohar 
Hi Robert, thank you for testing. I use mbed and perhaps there is a problem there. I will look into it.

I use mbed to create projects. I am newbie with ARM tooling. How can I create a project with this package (without mbed)?

**Comment by JonatanAntoni** [comment]
@gerhil this depends on the toolchain and ide you are using. The basic idea of CMSIS software packs is described in our [CMSIS documentation](http://www.keil.com/pack/doc/CMSIS/Pack/html/index.html). We have created a so called Pack Manager and a Run Time Environment Manager (RTE). This is part of our MDK Development Tools, and provided as a open-source [plug-in for Eclipse](https://github.com/ARM-software/cmsis-pack-eclipse).

**Comment by gerhil** [comment]
Can the plug in be used with MCUXpresso? I use MCUXpresso and mbed at the moment.

**Comment by JonatanAntoni** [comment]
I have no further information here because MCUXpresso is a ecosystem tool. Refer to the MCUXpresso community as well, please.

**Comment by gerhil** [comment]
How can I install the plugin? The readme says "Download the CMSISPackPlugin_x.x.zip from corresponding release", but where can I find the file?

**Comment by ilg-ul** [comment]
> Can the plug in be used with XXX? 

the general answer is 'yes, as long the XXX tools integrate the ARM packs plug-ins', otherwise expect things to be difficult (this is similar to the classical 'can I have my Ford in any colour? yes, as long as it is black').

I did not check the latest version, but all my previous attempts to use the plug-ins failed, for various reasons, and I finally gave up.

of course, if you have better experiences, or if you find the plug-ins integrated in other tools, please let me know, to re-evaluate.

**Comment by gerhil** [comment]
@RobertRostohar I have updated mbed but still the same output. The printf statements show weird output, but it is alternating between Taskl1 and Task2. So, I have surrounded the printf statements with osKernelLock() and osKernelUnlock(). I get good printf output but the example is not alternating anymore. The output is:
```
Task1
Task1
Task1
Task1
```
etc.
Any idea what the problem is?

**Comment by RobertRostohar** [comment]
@gerhil 
Let me clarify my previous statements. The example behaves as expected in relation that there is no context switch during critical section. However it doesn't work in a way that you have desired.

I fully agree with following statements from this thread.

> the semantics of scheduler critical sections is very simple and clear: 'no context switches are performed inside the critical section

>  if you call resume(), the thread will be added to the ready list, but, being inside the scheduler critical section, the context switch is ignored for now. 

> if you call suspend(), normally the current thread is suspended and execution continues to a thread chosen by the scheduler, 

> but, in a scheduler critical section, the current thread is resumed immediately 

```
void task1(void *parameter) {
    tid_thread1 = osThreadGetId();
    osThreadSuspend(tid_thread1);
    while(1) {
        printf ("Task1\r\n");
        fflush(stdout);
        osKernelLock();
        osThreadResume(tid_thread2);
        osThreadSuspend(tid_thread1);
        osKernelUnlock();
    }
}
```
- osKernelLock() enters critical section
- osThreadResume(tid_thread2) puts thread2 into ready list however does not switch to it since scheduler is locked
- osThreadSuspend(tid_thread1) would normally suspend running thread1 however since scheduler is locked it returns immediately without context switch
- osKernelUnock() exits critical section
- since both running thread1 and ready thread2 have the same priority execution continues with thread1 and repeats the above steps until round robin switches to thread2 (outside critical section)

The described behaviour is expected and explains also the disordered printf outputs.

**Comment by ilg-ul** [comment]
> ... I fully agree with following statements ...

ok, great.

but what do you think about the following use case:

```C++
void task1(void *parameter) {
    tid_thread1 = osThreadGetId();
    osThreadSuspend(tid_thread1);
    while(1) {
        osKernelLock();
        printf ("Task1\r\n");
        fflush(stdout);
        osKernelUnlock();

        osThreadResume(tid_thread2);
     }
}
```

except possible `printf()` oddities, in my opinion it should work. any idea why it did not in @gerhil's test?

**Comment by ilg-ul** [comment]
> there is also no suspend immediately 

`resume()`, if not in a critical section, should be immediate, i.e. suspend the current thread.

**Comment by RobertRostohar** [comment]
`resume()` will put thread2 into ready list however it won't pre-empt running thread1 (both have same priority). Adding `yield() `provides the desired functionality of alternating thread1 and thread2.

```
void task1(void *parameter) {
    tid_thread1 = osThreadGetId();
    osThreadSuspend(tid_thread1);
    while(1) {
        osKernelLock();
        printf ("Task1\r\n");
        fflush(stdout);
        osKernelUnlock();

        osThreadResume(tid_thread2);
        osThreadYield();
     }
}
```

**Comment by RobertRostohar** [comment]
@gerhil
> The osThreadSuspend does not suspend the task. This is not what I would expect, but is this the right behavior?

Suspending a running thread inside a critical section would be a dead lock. Therefore the thread is effectively not suspended in that case.

See my previous comment an for example of alternating thread1 and thread2.

**Comment by ilg-ul** [comment]
> however it won't pre-empt running thread1 (both have same priority)

hmmm... I know that you may consider this a matter of personal preferences, but for me this is just another inconsistence.

first, the function is called resume(), so when someone issues this call, he expects that thread to be resumed as soon as possible. 

second, generally most/all system calls should also act as yields. (the reason is the same as for adding preemption to a cooperative scheduler, to increase scheduler turn around).

 so, both because this is a system call and because the own meaning of the 'resume' word, the `resume()` call is expected to yield immediately.

**Comment by gerhil** [comment]
@RobertRostohar 

> Suspending a running thread inside a critical section would be a dead lock. Therefore the thread is effectively not suspended in that case.
> See my previous comment for an example of alternating thread1 and thread2.

I had implemented the same example as you did and this works. In my opinion it is logical if the yield is part of the outer osKernelUnlock(). However, i still have an issue with Waza (my library) and the suspend()/Resume()/Lock mechanism. I am trying to figure out what that is.

**Comment by gerhil** [comment]
> however it won't pre-empt running thread1 (both have same priority)

This is correct.

> resume(), if not in a critical section, should be immediate, i.e. suspend the current thread.

Ilg-ul, you have mentioned this before and I have been thinking about it. This assumption would be formally incorrect! I would not rely on such assumption. 

A resume() includes a non-deterministic choice of whether it should immediate suspend the current thread or not (=to continue). Any decision would be a valid decision. There is also a non-deterministic choice of which ready thread (or process) should run; the thread ready queue is non-deterministic. This fits the formal behavior of Parallel (CSP). This semantics also covers hardware parallellism where both threads continue at the same time by two parallel cores. This is valid for equal priorities. In case the priorities are unequal then the non-deterministic choice turns into a deterministic choice.

In case resume() suspends immeditely, should fits in this semantics, but I would not rely on it. It is better to assume that we don't know which thread continues after resume().

This is the same for threads being blocked and released on semaphores and monitors. But there is a concern that makes semaphores and monitors slighly behave different (without breaking the semantics). This concern is fairness, which is based on the shared resource being synchronized. Here must assume (according to Tony Hoare) that a release involves priorities of first-come-first-served on the waiting/condition/urgent queues.

**Comment by ilg-ul** [comment]
> > ... to increase scheduler turn around).
> No, not at all! 

what is wrong with increasing the scheduler turn around? is it useless or is it harmful? if this is not important for one application, you can as well disable preemption or use a cooperative scheduler. 

globally the system should not be affected, just that responsiveness may increase. you can achieve a similar effect by increasing the preemption clock frequency.

**Comment by gerhil** [comment]
@ilg-ul 
With preemption, the preemptive event should perform yield. This can be the end of a time-slice or a wakeup of a process.

Yielding is not harmful, but often unnecessary. Events (=rendezvous between processes) are the perfect moments to yield. Multithreading does not support this model. System calls that are blocking must yield. I avoid blocking system calls as much as possible. I use unbuffered channels to invoke these system calls. Channels are event makers. They determine when the program should yield (based on the event) and they determine which process should be resumed (based on engagement of the processes).

Events add responsive and reactive behavior to the program. Your design, i.e. the composition of processes and channels, will determine the optimal responsiveness and reactiveness of your program. There is no need to design with threads.

This is a perfect programming model to create durable software that can effectively use the energy functions of the ARM processor. Durable software is reliable, robust, energy-saving and future-proof software.

**Comment by ilg-ul** [comment]
@gerhil, I do not challenge your model, it may be superior, but we are currently operating within the limits of a multithreading RTOS, and you complained loudly against increasing the scheduler turn around, but later, when asked why, did not provide any reason (or I did not understand your reply).

I also understand that the behaviour of `resume()` may be non-determintistic, and applications should not count on it. ok, but if so, this must be clearly documented. however, since actual implementations usually make a decision on what to do, my recommendation for `resume()` is to yield, if not for other theoretical reason, at least for common sense consistency reasons and for improving responsiveness.

**Comment by gerhil** [comment]
> but we are currently operating within the limits of a multithreading RTOS, and you complained loudly against increasing the scheduler turn around, but later, when asked why, did not provide any reason

I use my previous reasoning in the model but also in the (non-CSP) real-time scheduler. I use synchronization/scheduling patterns that makes the scheduler yield only when necessary. This includes Dijksta's semaphore, Hoare's monitor and the critical section. I use the critical section to implementation of Dijkstra's semaphore, I use this to implement Hoare's monitor, and I use this to implement channels and the special process constructs (sequential, parallel, choice, catch). This work shows a balanced and optimal scheduler turn around that is defined by your design. Your design will determine the performance (responsiveness and reactiveness) of the programm. 

The exit of the critical section is important. It performs yield only when the thread administration has changed to a point where yield is necessary, i.e. when there is a new current thread.

My suspend function would look like this

```
void osThreadSuspendX(osThreadId_t t) {
  osKernelLock();    // <-- does not yield
  osThreadSuspend(t) // <-- does not yield, just administrative work
  osKernelUnlock();  // <-- performs yield when t is current thread (on new current thread)
}
```

My resume function would look like this

```
void osThreadResumeX(osThreadId_t t) {
  osKernelLock();   // <-- does not yield
  osThreadResume(t) // <-- does not yield, just administrative work
  osKernelUnlock(); // <-- performs yield on preemption, i.e. ...
  // ... only when a new higher priority thread gets in front of ready queue.
}
```

So when I use a nested critical section construct like this, the yield is pending until the exit of the critical section.

```
osKernelLock();      // <-- does not yield
osKernelLock();      // <-- does not yield
osThreadResumeX(x);  // <-- does not yield
osThreadSuspendX(y); // <-- does not yield
osKernelUnlock();    // <-- does not yield
osKernelUnlock();    // <-- performs yield when y is current thread or x has higher priority (on new current thread)
```

or

```
osKernelLock();     // <-- does not yield
osThreadResumeX(x); // <-- does not yield
osThreadResumeX(y); // <-- does not yield 
osThreadResumeX(s); // <-- does not yield
osThreadResumeX(t); // <-- does not yield
osKernelUnlock();  <-- performs yield only when necessary (on new current thread)
```

**Comment by ilg-ul** [comment]
> increasing the scheduler turn around ... I use my previous reasoning

ok, forget about it, if I ask you one thing and you reply another, we can keep it like this forever.

> a nested critical section

```C
// ...
osKernelUnlock();    // <-- does not yield
osKernelUnlock();    // <-- performs yield when y is current thread or x has higher priority (on new current thread)
```

nope. the first `osKernelUnlock()` will do exactly what it says, will unlock.

for nested critical sections, you need to save the result of `osKernelLock()` and to use it with `osKernelRestoreLock()`

the safe way to implement critical sections is:

```C
{
  int32_t state = osKernelLock();
  // ...
  osKernelRestoreLock(state);
}
```

this pattern works regardless the nesting level.

similarly you can implement un-critical sections (mirror images of critical sections, where you want the scheduler unlocked regardless its current status) with:

```C
{
  int32_t state = osKernelUnlock();
  // ...
  osKernelRestoreLock(state);
}
```

this is the C equivalent implementation of the C++ RAII (Resource Acquisition Is Initialization) pattern, with the resource acquired in the constructor and released in the destructor.

**Comment by gerhil** [comment]
> for nested critical sections, you need to save the result of osKernelLock() and to use it with osKernelRestoreLock()

You are right. I forgot about it. In Waza, I use `state` in the functions.

**Comment by gerhil** [comment]
@RobertRostohar 
> resume() will put thread2 into ready list however it won't pre-empt running thread1 (both have same priority). Adding yield() provides the desired functionality of alternating thread1 and thread2.

Your example is not useful. It can happen that some other process performs yield and the alternating behavior is gone.

I replied:

> I had implemented the an example with yield at the end and this works. 

I was wrong. The suspended threads play part of scheduling. This behavior is wrong.

Try

```
void task1(void *parameter) {
    tid_thread1 = osThreadGetId();
    osThreadSuspend(tid_thread1);
    while(1) {
        osKernelLock();
        printf ("Task1\r\n");
        fflush(stdout);
        osKernelUnlock();
        osKernelLock();
        osThreadResume(tid_thread2);
        osThreadSuspend(tid_thread1);
        osKernelUnlock();
        osThreadYield();
        osThreadYield(); // <-- extra yield
     }
}

void task2(void *parameter) {
    tid_thread2 = osThreadGetId();
    osThreadSuspend(tid_thread1);
    while(1) {
        osKernelLock();
        printf ("Task2\r\n");
        fflush(stdout);
        osKernelUnlock();
        osKernelLock();
        osThreadResume(tid_thread1);
        ofThreadSuspend(tid_thread2);
        osKernelUnlock();
        osThreadYield();
     }
}
```

Without the extra yield it seemed to work. But I was wrong. The extra yield should not change the alternation of both threads.

**Comment by gerhil** [comment]
@JonatanAntoni
@ReinhardKeil 
@RobertRostohar 

I would like to report this problem as a high priority BUG.

I am sure that I have shown in the example above than the osThreadSuspend() behaves incorrectly in the scope of a critical section (osKernelLock()/osKernelUnlock()). See previous post,

Note: I could not test this branch (cmsis_5) with the Gnu ARM tools, because I could not find the eclipse plug in. Instead, I have used the mbed branch (featured_cmsis5) and this works fine. I have used the latest updates.

FYI: I have found a similar problem with the ESP32 microsprocessor. Espressif has fixed the problem by performing a yield on a trigger when the interrupts are enabled. This is a very nice fix. I advise ARM to do something similar in the osKernelUnlock() function.

Suggestion: Why not renaming osKernelLock() and osKernelUnlock() to osCriticalSectionEnter() and osCriticalSectionExit()? The terms Lock and Unlock are confusing.

**Comment by ilg-ul** [comment]
> Why not renaming osKernelLock() and osKernelUnlock() to osCriticalSectionEnter() and osCriticalSectionExit()?

because they do not have identical semantic. and there can also be interrupt critical sections.

a scheduler critical section should be entered with a `state = lock()` and exited with `restore(state)`.
a scheduler uncritical section should be entered with a `state = unlock()` and exited with `restore(state)`.

**Comment by gerhil** [comment]
> a scheduler critical section should be entered with a state = lock() and exited with restore(state).

Yes. I have seen examples using osKernelLock(); ... o; osKernelUnlock(). I have used state-lock() and restore(state) and same problem.

I will check again.

**Comment by ilg-ul** [comment]
> performing a yield ... I advise ARM to do something similar in the osKernelUnlock() function.

you loudly complained about increasing scheduler turn around (and improving responsiveness) by performing implicit yields in system calls.

personally I do support this, yields should be performed on **all** system calls (of course, except those called from ISRs), and threads of equal priority should be given the CPU immediately, or right after the scheduler lock is released.

**Comment by ilg-ul** [comment]
> Yes. I have seen examples using osKernelLock(); ...  osKernelUnlock().

then they must be corrected. 

the save/restore pattern is just a good example of defensive programming. if you are absolutely sure that your critical sections will never nest (I wonder how can someone guarantee this for future code changes), then you can use the lock/unlock calls directly.

**Comment by gerhil** [comment]
> the save/restore pattern is just a good example of defensive programming. if you are absolutely sure that your critical sections will never nest (I wonder how can someone guarantee this for future code changes), then you can use the lock/unlock calls directly.

I agree. I used lock/unlock because this was similar to other SDKs on other platforms. 

I have updated the example and the result is the same. I have put the printf statements in the critical section for better output.

```
osThreadId_t tid_thread1;
osThreadId_t tid_thread2;

void task1(void *parameter) {
    tid_thread1 = osThreadGetId();
    osThreadSuspend(tid_thread1);
    while(1) {
        uint32_t state = osKernelLock();
        printf ("Task1\r\n");
        fflush(stdout);
        osThreadResume(tid_thread2);
        osThreadSuspend(tid_thread1);
        osKernelRestoreLock(state);
    }
}

void task2(void *parameter) {
    tid_thread2 = osThreadGetId();
    osThreadSuspend(tid_thread2);
    while(1) {
        uint32_t state = osKernelLock();
        printf ("Task2\r\n");
        fflush(stdout);
        osThreadResume(tid_thread1);
        osThreadSuspend(tid_thread2);
        osKernelRestoreLock(state);
    }
}

int main() {
    os_thread_t cb_thread1;
    memset(&cb_thread1, 0, sizeof(cb_thread1));
    osThreadAttr_t thread_attr1;
    memset(&thread_attr1, 0, sizeof(thread_attr1));
    thread_attr1.name = "thread1";
    thread_attr1.stack_size = 72U + (256U*4U);
    thread_attr1.stack_mem = new uint8_t[thread_attr1.stack_size];
    thread_attr1.priority = osPriorityNormal;
    thread_attr1.cb_mem = &cb_thread1;
    thread_attr1.cb_size = sizeof(os_thread_t);

    os_thread_t cb_thread2;
    memset(&cb_thread2, 0, sizeof(cb_thread2));
    osThreadAttr_t thread_attr2;
    memset(&thread_attr2, 0, sizeof(thread_attr2));
    thread_attr2.name = "thread2";
    thread_attr2.stack_size = 72U + (256U*4U);
    thread_attr2.stack_mem = new uint8_t[thread_attr2.stack_size];
    thread_attr2.priority = osPriorityNormal;
    thread_attr2.cb_mem = &cb_thread2;
    thread_attr2.cb_size = sizeof(os_thread_t);

    tid_thread1 = osThreadNew((osThreadFunc_t)task1, NULL, &thread_attr1);
    if (tid_thread1 == NULL) {
        printf("Thread1 not created!\r\n");
    }

    tid_thread2 = osThreadNew((osThreadFunc_t)task2, NULL, &thread_attr2);
    if (tid_thread2 == NULL) {
        printf("Thread2 not created!\r\n");
    }
    osDelay(1000); // Make sure both tasks are executed and suspended.
    osThreadResume(tid_thread1); // Start Task1 first.

for(;;) { osDelay(1000); }
return 0;
}
```

**Comment by ilg-ul** [comment]
as I already said, in my opinion, with normal implicit yield() behaviour, the `suspend()` should not be needed, and the code might read:

```C
void task1(void *parameter) {
    tid_thread1 = osThreadGetId();
    osThreadSuspend(tid_thread1);
    while (1) {
        { 
          // Scheduler critical section enter.
          uint32_t state = osKernelLock();

          printf ("Task1\r\n");
          fflush(stdout);

          osKernelRestoreLock(state);
          // Scheduler critical section exit.
        }
        osThreadResume(tid_thread2);
    }
}
```

if this does not work, then the RTOS implementation might have problems.

**Comment by gerhil** [comment]
This does not work as I expected. The osThreadSuspend is essential. The example is simple. One task must suspend and the other task must resume. This alternating behavior should not be influenced by osThreadYield() elsewhere in the program.

**Comment by ilg-ul** [comment]
> The osThreadSuspend is essential. 

it should not be

> The example is simple. One task must suspend and the other task must resume.

here we go again.

please understand that `resume()` is not called `enqueue_thread_to_ready_list()`, but `resume()` for good reasons, it should _resume_ the new thread, by doing an implicit `yield()` if there are threads with the same priority or higher, and the `yield()` includes and implicit `suspend()`.

`suspend()` is intended to implement synchronisation mechanisms, when there is nothing else to do in the current thread (and I insist it should not be allowed to be called with other threads), like waiting on a semaphore. 

`resume()` has a slightly more complex semantic, since it may be called from ISRs or from threads:

- when called from ISRs, it adds the given thread to the ready list, but cannot do more, since it is in the ISR context; for example releasing a semaphore from ISR. 
- when called from a thread, in addition to adding the thread to the ready list, it yields, to give the new thread a chance to run (if priorities are right); for example releasing a semaphore from another thread.

adding an explicit `suspend()` after `resume()` is acceptable as long as this is what you want, since it means that after returning from the `suspend()` included in the yielding`resume()`, which was triggered by another `resume()` called from another ISR/thread, the thread will suspend itself again, waiting for a second `resume()`.

if the current behaviour is different, then indeed we have a bug.

**Comment by gerhil** [comment]
I understand what you are saying. The example changes things for the resume and suspend. The example does not work. This example works for FreeRTOS, Java, Linux and Windows.

**Comment by JonatanAntoni** [comment]
Hello @gerhil,

I slightly modified your reproducer from above to demonstrate the current behavior:

```
#include <stdio.h>
#include "cmsis_os2.h"                  // ARM::CMSIS:RTOS2:Keil RTX5

static osThreadId_t tid_thread1;
static osThreadId_t tid_thread2;

static __attribute__((noreturn)) void task1(void *parameter);
static __attribute__((noreturn)) void task2(void *parameter);

void task1(void *parameter __attribute__((unused))) {
	osThreadYield();
	while(1) {
		int32_t state = osKernelLock();
		printf ("Task1: ");
		if (osThreadResume(tid_thread2) != osOK) {
			printf ("not resumed! ");
		} else {
			printf ("resumed. ");
		}
		if(osThreadSuspend(tid_thread1) != osOK) {
			printf ("not suspended! ");
		} else {
			printf ("suspended. ");
		}
		printf ("\r\n");
		fflush(stdout);
		osKernelRestoreLock(state);
		osThreadYield();
	}
}

void task2(void *parameter __attribute__((unused))) {
	osThreadSuspend(tid_thread2);
	while(1) {
		int32_t state = osKernelLock();
		printf ("Task2: ");
		if (osThreadResume(tid_thread1) != osOK) {
			printf ("not resumed! ");
		} else {
			printf ("resumed. ");
		}        
		if (osThreadSuspend(tid_thread2) != osOK) {
			printf ("not suspended! ");
		} else {
			printf ("suspended. ");
		}
		printf ("\r\n");
		fflush(stdout);
		osKernelRestoreLock(state);
		osThreadYield();
	}
}

int main() {
  osKernelInitialize();                 // Initialize CMSIS-RTOS
  
  tid_thread1 = osThreadNew(task1, NULL, NULL);
  if (tid_thread1 == NULL) {
		printf("Thread1 not created!\r\n");
  }

  tid_thread2 = osThreadNew(task2, NULL, NULL);
  if (tid_thread2 == NULL) {
		printf("Thread2 not created!\r\n");
  }
    
  if (osKernelGetState() == osKernelReady) {
    osKernelStart();                    // Start thread execution
  }
    
  while(1);
}
```
It gives me:
```
Task1: resumed. not suspended! 
Task2: not resumed! not suspended! 
Task1: not resumed! not suspended! 
Task2: not resumed! not suspended! 
Task1: not resumed! not suspended! 
```

osThreadSuspend() returns an error when called to suspend the running thread with scheduler locked. osThreadSuspend() is not meant to return until one calls osThreadResume() is called for the target thread. Subsequent calls to osThreadResume() to a not suspended thread returns errors as well.

The altering between both threads are due to the osThreadYield() calls I added. These force a round-robin-schedule even if the running thread has not exceeded its time slice.

As far as I understood you rather expect that for:
```
int32_t state = osKernelLock();
osThreadResume(tid_thread1);
osThreadSuspend(tid_thread2);
osKernelRestoreLock(state);
```

the call to osThreadSuspend() should return. Thus the thread is not suspended instantly but as soon as the scheduler gets unlocked? This means osThreadSuspend() should get an exceptional behavior if its called in a scheduler critical region, instead of returning an error?

**Comment by gerhil** [comment]
@JonatanAntoni 
Please add the extra yield. This extra yield should have no influence,

**Comment by gerhil** [comment]
@JonatanAntoni
Thanks for updating the code. The startup suspend/resume in my example was not necessary. I have added them to trace the behavior at the startup.

**Comment by JonatanAntoni** [comment]
@gerhil to visualize the impact of an extra yield, I duplicated the osThreadYield() only in task1. Now I get
```
Task1: resumed. not suspended! 
Task2: not resumed! not suspended! 
Task2: not resumed! not suspended! 
Task1: not resumed! not suspended! 
Task2: not resumed! not suspended! 
Task2: not resumed! not suspended! 
Task1: not resumed! not suspended! 
```
This is because both threads are ready all the time. Thus a yield forces a thread switch. Task one yields twice as task two.

Using FreeRTOS I can confirm a different behavior:
```
Task1: resumed. suspended. 
Task2: resumed. suspended. 
Task1: resumed. suspended. 
Task2: resumed. suspended. 
Task1: resumed. suspended. 
task1: resumed. suspended.
```
In this case the calls to yield does not have any impact because only one thread is ready at a time.

**Comment by gerhil** [comment]
@JonatanAntoni 

Great! Thanks for comparing it with FreeRTOS. Now we see a difference. What do you think? Should cmsis behave similar as FreeRTOS in this example? I would vote yes.

**Comment by JonatanAntoni** [comment]
We have to think about this carefully, so please be patient.
Are you blocked on this? Or are you able to continue your work based on CMSIS-FreeRTOS for now?

**Comment by JonatanAntoni** [comment]
@gerhil can you please enable `configASSERT` on your FreeRTOS example?

I did this as I am analyzing the issue. Calling vTaskSuspend when scheduler is locked leads to an assertion for me. Referring to the FreeRTOS documentation of [vTaskSuspendAll](http://www.freertos.org/a00134.html) reveals that one **must not** call `functions that have the potential to cause a context switch [..] while the scheduler is suspended`.

**Comment by gerhil** [comment]
@JonatanAntoni Interesting. I cannot use CMSIS-FreeRTOS directly because I have no Keil MDK license. I will try it with MCUXpresso. BTW I am using taskENTER_CRITICAL ()/taskEXIT_CRITICAL (). The critical sections I use are ultra small and ISRs of blocking I/O can resume threads.

**Comment by gerhil** [comment]
@JonatanAntoni No problems with taskENTER_CRITICAL ()/taskEXIT_CRITICAL () and vTaskSuspend()/vTaskResume().

**Comment by JonatanAntoni** [comment]
[CMSIS-FreeRTOS](https://github.com/ARM-software/CMSIS-FreeRTOS) is completely free and open source, no need for a Keil MDK license. It is a simple CMSIS-RTOS2 wrapper around vanilla FreeRTOS. Thus makes it very simple to switch back and forth between RTX5 an FreeRTOS.

If you do not want to use MDK with RTE, you might also manually include the CMSIS-Packs to your project. CMSIS-Packs are plain zip archives, beside the file extension and an additional pack description file inside.

**Comment by gerhil** [comment]
I would love to use MDK and RTE, but it's not for free, is it?

**Comment by JonatanAntoni** [comment]
You may start with [MDK-Lite](http://www2.keil.com/mdk5/editions/lite), this is free but has a 32kb code size limit for the linker.

**Comment by gerhil** [comment]
Hi @JonatanAntoni 

I have installed MDK-Lite and I can compile the project, but I cannot print the output of the FDRM-K64F in the output window. printf prevents the program to run. I have tried to retarget the stdout, but I don't know how to implement stdout_putchar() to output to the uart. 

I hope that this discussion pinpoints the problem and I hope that this will be solved in the near future. It would be great to use cmsis5 as an alternative to FreeRTOS.

Cheers, Gerald.

**Comment by JonatanAntoni** [comment]
@gerhil please let us fork this tool support issue into a separate discussion, i.e. per mail. Preferably we remove those off-topic posts afterwards to cleanup this thread in favor of the original issue.

**Comment by gerhil** [comment]
@JonatanAntoni 
Oke

**Comment by JonatanAntoni** [comment]
Hi @gerhil, 

do you still have any concerns about suspending tasks?

I reviewed the code of FreeRTOS once more. I use release 9.0.0 and my implementation of vTaskSuspend states at line 1618:
```
if( pxTCB == pxCurrentTCB )
{
	if( xSchedulerRunning != pdFALSE )
	{
		/* The current task has just been suspended. */
		configASSERT( uxSchedulerSuspended == 0 );
		portYIELD_WITHIN_API();
	}
	else
	[..]
}
```
This if statements cover the case when you suspend the current task after the scheduler has been started. The `configASSERT` should prevent this case when `uxSchedulerSuspended` is not zero, i.e. the scheduler was locked by a call to `vTaskSuspendAll`.

**Comment by gerhil** [comment]
Hi @JonatanAntoni 

I can't use CMSIS5 for my projects because the suspend and resume functions do not work well in critical sections. This needs to be fixed by ARM. For now I put CMSIS5 aside.

Instead, I use FreeRTOS with MCUXpresso. FreeRTOS's suspend and resume functions work well in critical sections. The portYIELD_WITHIN_API() functions in vTaskSuspend () and vTaskResume() do not yield in critical sections. This behavior is what I would expect from suspend (and resume) in CMSIS5, but CMSIS5 fails.

I love to use CMSIS5 with Waza (for C++) in the near future. Waza is a true concurrent programming model useful for embedded processors. Waza takes CMSIS5 to a much higher quality standard. It makes parallel software sound, simple and defect-free on ARM processors. Multithreading API’s can’t do this!

Currently, I am researching new design patterns for developing energy-saving software. This is part of Waza. Using concurrency is the most effective way to save energy in software. Unfortunately, I can’t demonstrate this with CMSIS5.

**Comment by JonatanAntoni** [comment]
Hi @gerhil,

Thanks for your valuable feedback by providing new use cases for CMSIS-RTOS. As I have to balance between many interests when changing existent APIs I might not be able to refine `osKernelLock()` and `osThreadSuspend()` to fit your needs.

Perhaps we can come up with an alternative solution that would work for you. As far as I understood you try to reuse some RTOS internals, i.e. the context switching. But you are not interested in the major parts of CMSIS-RTOSv2 public API. You are looking for functions that allow you to manipulate the ready list of the underlying RTOS and trigger context switching somehow explicitly. Is this correct? 

Are you interested in an implementation agnostic API, i.e. to be able to put Waza on top of any RTOS implementation available? Or do you mind to integrate RTX closely?

**Comment by gerhil** [comment]
Hi @JonatanAntoni 

IMO, you don’t have to change the API definitions to fix the problem. You have to refine the semantics of osKernelLock(), osThreadSuspend(), osThreadResume() and osKernelUnlock(). Nothing will break and nothing will be out of balance. The best option is fixing this problem. 

Perhaps 4 new functions could help me, something like:

osCriticalSectionBegin(); // like osKernelLock() (remember current thread)
osThreadSuspending(); // like osThreadSuspend() without yielding
osThreadResuming(); // like osThreadResume() without yielding
osCriticalSectionEnd(); // like osKernelUnlock() or restore with yielding (yields when the current thread has been changed into another thread)

These functions can help to improve CMSIS.. These functions are enough to get Waza working.

**Comment by JonatanAntoni** [comment]
Hi @gerhil, thank you for your enhancement proposals. We will need some time to clearly think about all the implications of either solution. Please be patient, I will report back as soon as possible.

**Comment by gerhil** [comment]
I understand. I am patient. I'm looking forward to hearing from you.
